### ISSUE TITLE
[Bug]: FourierRotaryEmbedding pads sin with 1 and corrupts unrotated channels

### ISSUE THREAD

**Issue by kvcache670** [vllm-project/vllm#55204]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
uv is set
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version                : 15.0.0 (git@github.com:llvm/llvm-project.git 4ba6a9c9f65bbc8bd06e3652cb20fd4dfc846137)
CMake version                : version 3.22.1
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.13.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.13 (main, Jun 23 2026, 15:18:55) [Clang 22.1.3 ] (64-bit runtime)
Python platform              : Linux-6.8.0-136-generic-x86_64-with-glibc2.35
    
==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   : 
GPU models and configuration : GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version        : 580.173.02
cuDNN version                : Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : False

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  24
On-line CPU(s) list:                     0-23
Vendor ID:                               GenuineIntel
Model name:                              13th Gen Intel(R) Core(TM) i7-13700KF
CPU family:                              6
Model:                                   183
Thread(s) per core:                      2
Core(s) per socket:                      16
Socket(s):                               1
Stepping:                                1
CPU max MHz:                             5800.0000
CPU min MHz:                             800.0000
BogoMIPS:                                6835.20
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               640 KiB (16 instances)
L1i cache:                               768 KiB (16 instances)
L2 cache:                                24 MiB (10 instances)
L3 cache:                         
```

</details>

Also verified on nightly 0.28.1rc1.dev315 on 2026-09-02. The reproduction requires no GPU, model weights, or server.

### 🐛 Describe the bug

`FourierRotaryEmbedding._compute_cos_sin_cache` pads both cosine and sine values to `head_size // 2` with `1`:

```python
sin = F.pad(input=sin, pad=(0, self.head_size // 2 - sin.size(-1)),
            mode="constant", value=1)
cos = F.pad(input=cos, pad=(0, self.head_size // 2 - cos.size(-1)),
            mode="constant", value=1)
```

Padding cosine with 1 is correct, but sine must be padded with 0 for the transformation to remain an identity. With both values set to 1, a padded channel becomes:

```text
out[j] = x[j] * 1 + rotate_neox(x)[j] * 1 = x[j] - x[j + head_size/2]
```

Instead of passing through unchanged, the channel becomes the difference between two unrelated channels.

This code path is used only by checkpoints with `rope_scaling.use_fope: true`. The incorrect padding is reached when fewer than `head_size // 2` frequencies remain, for example with a partial rotary factor, a smaller `num_inv_freq`, or a short `max_position_embeddings` that removes frequencies through the `inv_freq > 2π / max_position_embeddings` filter. FoPE checkpoints that require no padding are unaffected.

No error is raised; the model silently computes incorrect activations.

### Reproduction

```python
# pip install vllm==0.28.0 — CPU only, no GPU/checkpoint needed
import torch
from vllm.config import VllmConfig, set_current_vllm_config
from vllm.config.device import DeviceConfig
from vllm.model_executor.layers.rotary_embedding.fope import FourierRotaryEmbedding

HEAD, ROT = 64, 32                       # partial_rotary_factor = 0.5

with set_current_vllm_config(VllmConfig(device_config=DeviceConfig(device="cpu"))):
    rope = FourierRotaryEmbedding(
        head_size=HEAD, rotary_dim=ROT, max_position_embeddings=128,
        base=10000.0, is_neox_style=True, dtype=torch.float32, init_cache=False,
        num_key_value_heads=1, num_inv_freq=ROT // 2,
        fope_sep_head=True, fope_init_factor=1.0)

n_freq = rope.input_dim
with torch.no_grad():                     # identity mixing coefficients
    rope.cos_coef.copy_(torch.eye(n_freq).unsqueeze(0))
    rope.sin_coef.copy_(torch.eye(n_freq).unsqueeze(0))
rope.update_cache = True

torch.manual_seed(0)
q = torch.randn(2, HEAD)
q_out, _ = rope.forward_native(torch.arange(2), q.clone(), torch.randn(2, HEAD).clone())
q_out = q_out.view(2, HEAD)

untouched = int((q_out == q).all(dim=0).sum().item())
print(f"channels bit-identical: {untouched}/{HEAD} (correct partial rotary: {HEAD - ROT})")
# prints 0/64 — not one channel survives, half of them should
```

### Expected behavior

Pad sine with 0 while keeping cosine padded with 1, so an unrotated channel remains `x[j] * 1 + rotate(x)[j] * 0 = x[j]`.

Separately, the same constructor crashes when `fope_sep_head=False` because the `einsum` uses two subscripts with a three-dimensional `cos_coef`.

_This issue was discovered by a fuzzer, and the report was co-authored with an LLM._

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by modelpath-dev** [comment]
I'll work on this. Could you assign it to me?

I'll reproduce from the report, fix the failure path, and add a test so it stays fixed.

**Comment by modelpath-dev** [comment]
I am still working on reproducing the issue. I will update once I have more information or if I need any assistance. Thank you for your patience.

**Comment by modelpath-dev** [comment]
I have managed to reproduce the issue. I am now working on a fix. I will update again once I have a solution or if I encounter any problems. Thank you for your patience.
